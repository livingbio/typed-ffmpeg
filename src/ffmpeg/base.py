from typing import Any

from ffmpeg.schema import StreamType

from .dag.nodes import FilterableStream, FilterNode, InputNode, MergeOutputsNode, OutputNode, OutputStream
from .streams.audio import AudioStream
from .streams.av import AVStream
from .streams.video import VideoStream


def input(filename: str, **kwargs: Any) -> AVStream:
    """
    Input file URL (ffmpeg ``-i`` option)

    Args:
        filename: Input file URL
        **kwargs: ffmpeg's input file options

    Returns:
        Input stream

    Examples:
    ```py
    >>> input('input.mp4')
    <AVStream:input.mp4:0>
    ```
    """
    return InputNode(filename=filename, kwargs=tuple(kwargs.items())).stream()


def output(*streams: FilterableStream, filename: str, **kwargs: Any) -> OutputStream:
    """
    Output the streams to a file URL

    Args:
        *streams: the streams to output
        filename: the filename to output to
        **kwargs: the arguments for the output

    Returns:
        the output stream
    """
    return OutputNode(filename=filename, inputs=streams, kwargs=tuple(kwargs.items())).stream()


def merge_outputs(*streams: OutputStream) -> OutputStream:
    """
    Merge multiple output streams into one.

    Args:
        *streams: The output streams to merge.

    Returns:
        The merged output stream.
    """
    return MergeOutputsNode(inputs=streams).stream()


def vfilter(
    *streams: FilterableStream, name: str, input_typings: tuple[StreamType, ...] = (StreamType.video,), **kwargs: Any
) -> VideoStream:
    """
    Apply a custom video filter which has only one output to this stream

    Args:
        *streams: the streams to apply the filter to
        name: the name of the filter
        **kwargs: the arguments for the filter

    Returns:
        the output stream

    Note:
        This function is for custom filter which is not implemented in typed-ffmpeg
    """
    return FilterNode(
        name=name,
        inputs=streams,
        output_typings=(StreamType.video,),
        input_typings=input_typings,
        kwargs=tuple(kwargs.items()),
    ).video(0)


def afilter(
    *streams: FilterableStream, name: str, input_typings: tuple[StreamType, ...] = (StreamType.audio,), **kwargs: Any
) -> AudioStream:
    """
    Apply a custom audio filter which has only one output to this stream

    Args:
        *streams: the streams to apply the filter to
        name: the name of the filter
        **kwargs: the arguments for the filter

    Returns:
        the output stream

    Note:
        This function is for custom filter which is not implemented in typed-ffmpeg
    """
    return FilterNode(
        name=name,
        inputs=streams,
        output_typings=(StreamType.audio,),
        input_typings=input_typings,
        kwargs=tuple(kwargs.items()),
    ).audio(0)


def filter_multi_output(
    *streams: FilterableStream,
    name: str,
    input_typings: tuple[StreamType, ...] = (),
    output_tyings: tuple[StreamType, ...] = (),
    **kwargs: Any
) -> FilterNode:
    """
    Apply a custom filter which has multiple outputs to this stream

    Args:
        *streams: the streams to apply the filter to
        name: the name of the filter
        **kwargs: the arguments for the filter

    Returns:
        the FilterNode

    Note:
        This function is for custom filter which is not implemented in typed-ffmpeg
    """
    return FilterNode(
        name=name,
        kwargs=tuple(kwargs.items()),
        inputs=streams,
        input_typings=input_typings,
        output_typings=output_tyings,
    )
