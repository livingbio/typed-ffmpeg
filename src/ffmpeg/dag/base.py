from __future__ import annotations

from abc import ABC, abstractmethod, abstractproperty
from dataclasses import dataclass
from functools import cached_property
from typing import Iterable, Sequence

from ..schema import StreamType
from ..utils.dag import is_dag
from ..utils.typing import override


class _DAGContext(ABC):
    """
    An abstract class context for DAG.
    """

    @abstractmethod
    def get_node_label(self, node: Node) -> str:
        """
        Get the label of the node.

        Args:
            node: The node to get the label of.

        Returns:
            The label of the node.
        """
        raise NotImplementedError()

    @abstractmethod
    def get_outgoing_streams(self, node: Node) -> Iterable[Stream]:
        """
        Extract all node's outgoing streams from the given set of streams, Because a node only know its incoming streams.

        Args:
            node: The node to get the outgoing streams of.

        Returns:
            The outgoing streams of the node.
        """
        raise NotImplementedError()


class DummyDAGContext(_DAGContext):
    """A dummy DAG context that does not do anything"""

    @override
    def get_node_label(self, node: Node) -> str:
        return str(node)

    @override
    def get_outgoing_streams(self, node: Node) -> Iterable[Stream]:
        return []


empty_dag_context = DummyDAGContext()


@dataclass(frozen=True, kw_only=True)
class HashableBaseModel:
    @cached_property
    def hex(self) -> str:
        return hex(abs(hash(self)))[2:]

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self.hex})"


@dataclass(frozen=True, kw_only=True)
class Stream(HashableBaseModel):
    """
    A 'Stream' represents a sequence of data flow in the Directed Acyclic Graph (DAG).

    Note:
        Each stream in the DAG is a sequence of operations that transforms the data from its input form to its output form. The stream is an essential component of the DAG, as it defines the order and the nature of the operations that are performed on the data.
    """

    node: Node
    """
    Represents the node that the stream is connected to in the upstream direction.

    Note:
        In the context of a data stream, the 'upstream' refers to the source of the data, or where the data is coming from. Therefore, the 'upstream node' is the node that is providing the data to the current stream.
    """
    selector: StreamType | None = None
    """
    Represents the type of the stream.

    Notes:
        See Also: [Stream specifiers](https://ffmpeg.org/ffmpeg.html#Stream-specifiers-1) `stream_type`
    """

    index: int | None = None  # the nth child of the node
    """
    Represents the index of the stream in the node's output streams.

    Note:
        See Also: [Stream specifiers](https://ffmpeg.org/ffmpeg.html#Stream-specifiers-1) `stream_index`
    """


@dataclass(frozen=True, kw_only=True)
class Node(HashableBaseModel, ABC):
    """
    A 'Node' represents a single operation in the Directed Acyclic Graph (DAG).

    Note:
        Each node in the DAG represents a single operation that transforms the data from its input form to its output form. The node is an essential component of the DAG, as it defines the nature of the operations that are performed on the data.
    """

    args: tuple[str, ...] = ()
    kwargs: tuple[tuple[str, str | int | float | bool], ...] = ()

    @abstractproperty
    def incoming_streams(self) -> Sequence["Stream"]:
        """
        Return the incoming streams of this node.

        Returns:
            The incoming streams of this node.
        """
        raise NotImplementedError()

    def __post_init__(self) -> None:
        # Validate the DAG
        passed = set()
        nodes = [self]
        output = {}

        while nodes:
            node = nodes.pop()

            if node in passed:
                continue
            passed.add(node)

            nodes.extend(k.node for k in node.incoming_streams)

            output[str(node)] = set(str(k.node) for k in node.incoming_streams)

        if not is_dag(output):
            raise ValueError(f"Graph is not a DAG: {output}")

    @abstractmethod
    def get_args(self, context: _DAGContext = empty_dag_context) -> list[str]:
        """
        Get the arguments of the node.

        Args:
            context: The DAG context. Defaults to empty_dag_context.

        Returns:
            The arguments of the node.
        """
        raise NotImplementedError()
