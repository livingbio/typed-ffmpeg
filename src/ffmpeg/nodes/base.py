from __future__ import annotations

from abc import ABC, abstractmethod, abstractproperty
from dataclasses import dataclass
from functools import cached_property
from typing import Any, Iterable, Sequence

from ..schema import StreamType
from ..utils.dag import is_dag


class _DAGContext(ABC):
    @abstractmethod
    def get_node_label(self, node: Node) -> str:
        """Get the label of the node"""
        raise NotImplementedError()

    @abstractmethod
    def get_outgoing_streams(self, node: Node) -> Iterable[Stream]:
        """Extract all node's outgoing streams from the given set of streams, Because a node only know its incoming streams"""
        raise NotImplementedError()


class DummyDAGContext(_DAGContext):
    """A dummy DAG context that does not do anything"""

    def get_node_label(self, node: Node) -> str:
        return str(node)

    def get_outgoing_streams(self, node: Node) -> Iterable[Stream]:
        return []


empty_dag_context = DummyDAGContext()


@dataclass(frozen=True, kw_only=True)
class HashableBaseModel:
    @cached_property
    def hex(self) -> str:
        return hex(abs(hash(self)))[2:]

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self.hex})"


@dataclass(frozen=True, kw_only=True)
class Stream(HashableBaseModel):
    node: Node
    selector: StreamType | None = None
    index: int | None = None  # the nth child of the node


@dataclass(frozen=True, kw_only=True)
class Node(HashableBaseModel, ABC):
    args: tuple[str, ...] = ()
    kwargs: tuple[tuple[str, Any], ...] = ()

    @abstractproperty
    def incoming_streams(self) -> Sequence["Stream"]:
        raise NotImplementedError()

    def __post_init__(self) -> None:
        passed = set()
        nodes = [self]
        output = {}

        while nodes:
            node = nodes.pop()

            if node in passed:
                continue
            passed.add(node)

            nodes.extend(k.node for k in node.incoming_streams)

            output[str(node)] = set(str(k.node) for k in node.incoming_streams)

        if not is_dag(output):
            raise ValueError(f"Graph is not a DAG: {output}")

    @abstractmethod
    def get_args(self, context: _DAGContext = empty_dag_context) -> list[str]:
        raise NotImplementedError()
