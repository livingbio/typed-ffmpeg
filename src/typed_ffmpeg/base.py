from .nodes.nodes import FilterableStream, InputNode, MergeOutputsNode, OutputNode, OutputStream
from .streams.av import AVStream


def input(filename: str, **kwargs: str | int | None | float) -> AVStream:
    """Input file URL (ffmpeg ``-i`` option)
    Any supplied kwargs are passed to ffmpeg verbatim (e.g. ``t=20``,
    ``f='mp4'``, ``acodec='pcm'``, etc.).
    To tell ffmpeg to read from stdin, use ``pipe:`` as the filename.
    Official documentation: `Main options <https://ffmpeg.org/ffmpeg.html#Main-options>`__
    """
    fmt = kwargs.pop("f", None)
    if fmt:
        if "format" in kwargs:
            raise ValueError("Can't specify both `format` and `f` kwargs")
        kwargs["format"] = fmt
    return InputNode(filename=filename, kwargs=tuple(kwargs.items())).stream()


def output(*inputs: FilterableStream, filename: str, **kwargs: str | int | None | float) -> OutputStream:
    return OutputNode(filename=filename, inputs=inputs, kwargs=tuple(kwargs.items())).stream()


def merge_outputs(*inputs: OutputStream) -> OutputStream:
    return MergeOutputsNode(inputs=inputs).stream()


__all__ = ["input", "output", "merge_outputs"]
