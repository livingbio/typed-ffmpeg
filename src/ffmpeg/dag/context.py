from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from functools import cached_property
from typing import Any, TypeVar

from ..utils.typing import override
from .nodes import FilterNode, InputNode
from .schema import Node, Stream

T = TypeVar("T")


def _remove_duplicates(seq: list[T]) -> list[T]:
    """
    Remove duplicates from a list while preserving order.

    Args:
        seq: The list to remove duplicates from.

    Returns:
        The list with duplicates removed.
    """
    seen = set()
    output = []

    for x in seq:
        if x not in seen:
            output.append(x)
            seen.add(x)

    return output


def _collect(node: Node) -> tuple[list[Node], list[Stream]]:
    """
    Collect all nodes and streams that are upstreamed to the given node

    Args:
        node: The node to collect from.

    Returns:
        A tuple of all nodes and streams that are upstreamed to the given node.
    """
    nodes, streams = [node], [*node.inputs]

    for stream in node.inputs:
        _nodes, _streams = _collect(stream.node)
        nodes += _nodes
        streams += _streams

    return nodes, streams


@dataclass(frozen=True, kw_only=True)
class DAGContext:
    """
    A context for a directed acyclic graph (DAG).
    """

    node: Node
    """
    The root node (the destination) of the DAG.
    """

    nodes: tuple[Node, ...]
    """
    All nodes in the graph.
    """

    streams: tuple[Stream, ...]
    """
    All streams in the graph.
    """

    @classmethod
    def build(cls, node: Node) -> DAGContext:
        """
        create a DAG context based on the given node

        Args:
            node: The root node of the DAG.

        Returns:
            A DAG context based on the given node.
        """
        nodes, streams = _collect(node)

        return cls(
            node=node,
            nodes=tuple(_remove_duplicates(nodes)),
            streams=tuple(_remove_duplicates(streams)),
        )

    @cached_property
    def all_nodes(self) -> list[Node]:
        """
        All nodes in the graph sorted by the number of upstream nodes.
        """
        return sorted(self.nodes, key=lambda node: len(node.upstream_nodes))

    @cached_property
    def all_streams(self) -> list[Stream]:
        """
        All streams in the graph sorted by the number of upstream nodes and the index of the stream.
        """
        return sorted(self.streams, key=lambda stream: (len(stream.node.upstream_nodes), stream.index))

    @cached_property
    def outgoing_nodes(self) -> dict[Stream, list[tuple[Node, int]]]:
        """
        A dictionary of outgoing nodes for each stream.
        """
        outgoing_nodes: dict[Stream, list[tuple[Node, int]]] = defaultdict(list)

        for node in self.nodes:
            for idx, stream in enumerate(node.inputs):
                outgoing_nodes[stream].append((node, idx))

        return outgoing_nodes

    @cached_property
    def outgoing_streams(self) -> dict[Node, list[Stream]]:
        """
        A dictionary of outgoing streams for each node.
        """

        outgoing_streams: dict[Node, list[Stream]] = defaultdict(list)

        for stream in self.streams:
            outgoing_streams[stream.node].append(stream)

        return outgoing_streams

    @cached_property
    def node_labels(self) -> dict[Node, str]:
        """
        A dictionary of outgoing streams for each node.
        """

        input_node_index = 0
        filter_node_index = 0
        node_labels: dict[Node, str] = {}

        for node in sorted(self.nodes, key=lambda node: node.max_depth):
            if isinstance(node, InputNode):
                node_labels[node] = str(input_node_index)
                input_node_index += 1
            elif isinstance(node, FilterNode):
                node_labels[node] = f"s{filter_node_index}"
                filter_node_index += 1
            else:
                node_labels[node] = "out"

        return node_labels

    @override
    def get_outgoing_nodes(self, stream: Stream) -> list[tuple[Node, int]]:
        """
        Get all outgoing nodes of the stream.

        Args:
            stream: The stream to get the outgoing nodes of.

        Returns:
            The outgoing nodes of the stream.
        """
        return self.outgoing_nodes[stream]

    @override
    def get_node_label(self, node: Node) -> str:
        """
        Get the label of the node.

        Args:
            node: The node to get the label of.

        Returns:
            The label of the node.
        """

        assert isinstance(node, (InputNode, FilterNode)), "Only input and filter nodes have labels"
        return self.node_labels[node]

    @override
    def get_outgoing_streams(self, node: Node) -> list[Stream]:
        """
        Extract all node's outgoing streams from the given set of streams, Because a node only know its incoming streams.

        Args:
            node: The node to get the outgoing streams of.

        Returns:
            The outgoing streams of the node.
        """
        return self.outgoing_streams[node]

    def render(self, obj: Any) -> Any:
        """
        Render the object to a string.

        Args:
            obj: The object to render.

        Returns:
            The rendered object.
        """

        if isinstance(obj, (list, tuple)):
            return [self.render(o) for o in obj]
        elif isinstance(obj, dict):
            return {self.render(k): self.render(v) for k, v in obj.items()}

        if isinstance(obj, Node):
            return f"Node({obj.repr()}#{self.node_labels[obj]})"

        if isinstance(obj, Stream):
            return f"Stream({self.render(obj.node)}#{obj.index})"

        return obj
