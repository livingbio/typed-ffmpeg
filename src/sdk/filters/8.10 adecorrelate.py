def adecorrelate(
    stream: Stream,
    stages: int = 6,
    seed: int = 0,
) -> Stream:
    """
    Apply decorrelation to input audio stream.

    The filter accepts the following options:

        Parameters:
        ----------
        stream : Stream
            The input stream to filter.
        stages : int, optional
            Set decorrelation stages of filtering. Allowed range is from 1 to 16.
            Default value is 6.
        seed : int, optional
            Set random seed used for setting delay in samples across channels.

    Example usage:
    --------------
    stream.adecorrelate(
        stages=8,
        seed=1234,
    )

    Ref: https://ffmpeg.org/ffmpeg-filters.html#adecorrelate
    """
    return FilterNode(
        stream,
        adecorrelate.__name__,
        kwargs={
            "stages": stages,
            "seed": seed,
        },
    ).stream()
