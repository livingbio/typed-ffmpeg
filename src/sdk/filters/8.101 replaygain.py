from typing import Optional

from ..node import FilterNode
from ..stream import Stream


def replaygain(stream: Stream, track_gain: Optional[str] = None, track_peak: Optional[str] = None) -> Stream:
    """
    ReplayGain scanner filter. This filter takes an audio stream as an input and outputs it unchanged.
    At end of filtering, it displays track_gain and track_peak.

    Parameters:
    ----------
    stream : Stream
        The input stream to filter.
    track_gain : str, optional
        Exported track gain in dB at the end of the stream.
    track_peak : str, optional
        Exported track peak at the end of the stream.

    Returns:
    -------
    Stream
        The filtered output stream.

    Example usage:
    --------------
    stream.replaygain(track_gain="track_gain.txt", track_peak="track_peak.txt")

    Ref: https://ffmpeg.org/ffmpeg-filters.html#replaygain
    """
    return FilterNode(
        stream,
        replaygain.__name__,
        kwargs={
            "track_gain": track_gain,
            "track_peak": track_peak,
        },
    ).stream()
