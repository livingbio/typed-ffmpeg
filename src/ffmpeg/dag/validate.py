from __future__ import annotations

from dataclasses import replace

from ffmpeg.exceptions import FFMpegValueError

from ..streams.audio import AudioStream
from ..streams.video import VideoStream
from .context import DAGContext
from .nodes import FilterNode, InputNode
from .schema import Node, Stream


def remove_split(current_stream: Stream, mapping: dict[Stream, Stream] = None) -> tuple[Stream, dict[Stream, Stream]]:
    """
    Rebuild the graph with the given mapping.

    Args:
        current_stream: The stream to rebuild the graph with.
        mapping: The mapping to rebuild the graph with.

    Returns:
        A tuple of the new node and the new mapping.
    """

    # remove all split nodes
    # add split nodes to the graph
    if mapping is None:
        mapping = {}

    if current_stream in mapping:
        return mapping[current_stream], mapping

    if not current_stream.node.inputs:
        mapping[current_stream] = current_stream
        return current_stream, mapping

    if isinstance(current_stream.node, FilterNode):
        # if the current node is a split node, we need to remove it
        if current_stream.node.name in ("split", "asplit"):
            new_stream, _mapping = remove_split(current_stream=current_stream.node.inputs[0], mapping=mapping)
            mapping[current_stream] = mapping[current_stream.node.inputs[0]]
            return mapping[current_stream.node.inputs[0]], mapping

    inputs = {}
    for idx, input_stream in sorted(
        enumerate(current_stream.node.inputs), key=lambda idx_stream: -len(idx_stream[1].node.upstream_nodes)
    ):
        new_stream, _mapping = remove_split(current_stream=input_stream, mapping=mapping)
        inputs[idx] = new_stream
        mapping |= _mapping

    new_node = replace(
        current_stream.node, inputs=tuple(stream for idx, stream in sorted(inputs.items(), key=lambda x: x[0]))
    )
    new_stream = replace(current_stream, node=new_node)

    mapping[current_stream] = new_stream
    return new_stream, mapping


def add_split(
    current_stream: Stream,
    down_node: Node = None,
    down_index: int = None,
    context: DAGContext = None,
    mapping: dict[tuple[Stream, Node | None, int | None], Stream] = None,
) -> tuple[Stream, dict[tuple[Stream, Node | None, int | None], Stream]]:
    """
    Add split nodes to the graph.

    Args:
        current_stream: The stream to add split nodes to.
        down_node: The node use current_stream as input.
        down_index: The index of the input stream in down_node.
        context: The DAG context.
        mapping: The mapping to add split nodes to.

    Returns:
        A tuple of the new node and the new mapping.
    """

    if not context:
        context = DAGContext.build(current_stream.node)

    if mapping is None:
        mapping = {}

    if (current_stream, down_node, down_index) in mapping:
        return mapping[(current_stream, down_node, down_index)], mapping

    inputs = {}

    for idx, input_stream in sorted(
        enumerate(current_stream.node.inputs), key=lambda idx_stream: -len(idx_stream[1].node.upstream_nodes)
    ):
        new_stream, _mapping = add_split(
            current_stream=input_stream, down_node=current_stream.node, down_index=idx, mapping=mapping, context=context
        )
        inputs[idx] = new_stream
        mapping |= _mapping

    new_node = replace(
        current_stream.node, inputs=tuple(stream for idx, stream in sorted(inputs.items(), key=lambda x: x[0]))
    )
    new_stream = replace(current_stream, node=new_node)

    num = len(context.get_outgoing_nodes(current_stream))
    if num < 2:
        mapping[(current_stream, down_node, down_index)] = new_stream
        return new_stream, mapping

    if isinstance(current_stream.node, InputNode):
        for idx, (node, index) in enumerate(context.get_outgoing_nodes(current_stream)):
            # if the current node is InputNode, we don't need to split it
            mapping[(current_stream, node, index)] = new_stream
        return new_stream, mapping

    if isinstance(new_stream, VideoStream):
        split_node = new_stream.split(outputs=num)
        for idx, (node, index) in enumerate(context.get_outgoing_nodes(current_stream)):
            mapping[(current_stream, node, index)] = split_node.video(idx)
        return mapping[(current_stream, down_node, down_index)], mapping
    elif isinstance(new_stream, AudioStream):
        split_node = new_stream.asplit(outputs=num)
        for idx, (node, index) in enumerate(context.get_outgoing_nodes(current_stream)):
            mapping[(current_stream, node, index)] = split_node.audio(idx)
        return mapping[(current_stream, down_node, down_index)], mapping
    else:
        raise FFMpegValueError(f"unsupported stream type: {current_stream}")


def fix_graph(stream: Stream) -> Stream:
    """
    Fix the graph by removing and adding split nodes.

    Args:
        stream: The stream to fix.

    Returns:
        The fixed stream.

    Note:
        Fix the graph by resetting split nodes.
        This function is for internal use only.
    """

    stream, _ = remove_split(stream)
    stream, _ = add_split(stream)
    return stream


def validate(stream: Stream, auto_fix: bool = True) -> Stream:
    """
    Validate the given DAG. If auto_fix is True, the graph will be automatically fixed to follow ffmpeg's rule.

    Args:
        stream: The DAG to validate.
        auto_fix: Whether to automatically fix the graph.

    Returns:
        The validated DAG context.
    """
    if auto_fix:
        stream = fix_graph(stream)

    # NOTE: we don't want to modify the original node
    # validators: list[] = []

    # for validator in validators:
    #     context = validator(context)

    return stream
