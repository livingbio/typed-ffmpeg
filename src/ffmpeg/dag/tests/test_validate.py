import pytest
from syrupy.assertion import SnapshotAssertion

from ...base import input
from ...filters import concat
from ..context import DAGContext
from ..validate import _validate_not_utilize_split, _validate_reuse_stream


def test_validate_reuse_stream(snapshot: SnapshotAssertion) -> None:
    input1 = input("input1.mp4")
    rev = input1.reverse()
    context = DAGContext.build(concat(rev.trim(), rev.trim()).video(0).output(filename="tmp.mp4"))

    with pytest.raises(AssertionError) as e:
        _validate_reuse_stream(context)

    assert snapshot == e


def test_validate_not_utilize_split(snapshot: SnapshotAssertion) -> None:
    input1 = input("input1.mp4")

    context = DAGContext.build(input1.split(outputs=2).video(0).output(filename="tmp.mp4"))
    with pytest.raises(AssertionError) as e:
        _validate_not_utilize_split(context)

    assert snapshot == e


def test_reduntant_split_outputs_1(snapshot: SnapshotAssertion) -> None:
    input1 = input("input1.mp4")

    context = DAGContext.build(input1.split(outputs=1).video(0).output(filename="tmp.mp4"))
    with pytest.raises(AssertionError) as e:
        _validate_not_utilize_split(context)

    assert snapshot == e


def test_reduntant_split_duplicate(snapshot: SnapshotAssertion) -> None:
    input1 = input("input1.mp4")
    s = input1.split(outputs=2)
    s0 = s.video(0)
    s1 = s.video(1)

    s00 = s0.split(outputs=2).video(0)
    s01 = s0.split(outputs=2).video(1)

    context = DAGContext.build(concat(s00, s01, s1, n=3).video(0).output(filename="tmp.mp4"))
    with pytest.raises(AssertionError) as e:
        _validate_not_utilize_split(context)

    assert snapshot == e
