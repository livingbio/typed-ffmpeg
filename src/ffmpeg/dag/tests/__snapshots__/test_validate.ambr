# serializer version: 1
# name: test_reduntant_split_duplicate
  <ExceptionInfo AssertionError("Found reduntant split nodes: {FilterNode(args=(), kwargs=(('outputs', 2),), inputs=(VideoStream(node=F...t_typings=(<StreamType.video: 'video'>,), output_typings=(<StreamType.video: 'video'>, <StreamType.video: 'video'>))}") tblen=2>
# ---
# name: test_reduntant_split_outputs_1
  <ExceptionInfo AssertionError("Found reduntant split nodes: {FilterNode(args=(), kwargs=(('outputs', 1),), inputs=(AVStream(node=Inpu...=None),), name='split', input_typings=(<StreamType.video: 'video'>,), output_typings=(<StreamType.video: 'video'>,))}") tblen=2>
# ---
# name: test_validate_not_utilize_split
  <ExceptionInfo AssertionError("Found not utilized split nodes: [FilterNode(args=(), kwargs=(('outputs', 2),), inputs=(AVStream(node=I...t_typings=(<StreamType.video: 'video'>,), output_typings=(<StreamType.video: 'video'>, <StreamType.video: 'video'>))]") tblen=2>
# ---
# name: test_validate_reuse_stream
  <ExceptionInfo AssertionError("Found reuse streams: [VideoStream(node=FilterNode(args=(), kwargs=(), inputs=(VideoStream(node=FilterN...nput_typings=(<StreamType.video: 'video'>,), output_typings=(<StreamType.video: 'video'>,)), selector=None, index=0)]") tblen=2>
# ---
